lib/cache.js

```diff
@@ -8,11 +8,29 @@ const VERSION = require('../package').version;
 
 const TimeoutError = bluebird.TimeoutError;
 
-function createCacheKey(segment, id) {
+function createCacheKey(segment, ctx, opts) {
   const versionedSegment = `http-transport:${VERSION}:${segment}`;
+  const requiredIdPart = `${ctx.req.getMethod()}:${ctx.req.getUrl()}`;
+  const varyOn = _.get(opts, 'varyOn');
+
+  if (!varyOn) {
+    return {
+      segment: versionedSegment,
+      id: requiredIdPart
+    };
+  }
+
+  const headers = ctx.req.getHeaders();
+  const optionalIdPart = varyOn.map((key) => {
+    if (headers[key]) {
+      return `${key}=${headers[key]}`;
+    }
+    return `${key}=`;
+  });
+
   return {
     segment: versionedSegment,
-    id
+    id: `${requiredIdPart}:${optionalIdPart.join(',')}`
```

Comment:
If there are no headers in the request, but a `varyOn` is supplied in the options, I think this would always append a `:` after the `requiredIdPart`. Is that intentional?

Reply:
If there are no headers in request and we provide `varyOn` option then I believe we'll get whatever we provide in the option plus `=`. For example

```
// option
const varyOn = [
  `accept-lang`,
  `accept`
];
```

`optionalIdPart` will be `accept-lang=,accept=`.

Reply:
Yes, that makes sense :)